File: crud-item-list/src/main/java/com/gurunars/crud_item_list/ActionEdit.java
package com.gurunars.crud_item_list;

import com.esotericsoftware.kryo.Kryo;

import org.objenesis.strategy.StdInstantiatorStrategy;

import java.util.Iterator;
import java.util.List;
import java.util.Set;

import java8.util.function.Consumer;

final class ActionEdit<ItemType> implements Action<ItemType> {

    private Kryo kryo = new Kryo();
    private Consumer<ItemType> itemConsumer;

    ActionEdit(Consumer<ItemType> itemConsumer) {
        this.kryo.setInstantiatorStrategy(new Kryo.DefaultInstantiatorStrategy(new StdInstantiatorStrategy()));
        this.itemConsumer = itemConsumer;
    }

    @Override
    public boolean perform(List<ItemType> all, Set<ItemType> selectedItems) {
        Iterator<ItemType> iterator = selectedItems.iterator();
        if (iterator.hasNext()) {
            itemConsumer.accept(kryo.copy(iterator.next()));
        }
        return false;
    }

    @Override
    public boolean canPerform(List<ItemType> all, Set<ItemType> selectedItems) {
        return selectedItems.size() == 1;
    }

}

File: crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/ActionEdit.kt
package com.gurunars.crud_item_list

import com.gurunars.box.IRoBox
import com.gurunars.box.oneWayBranch
import com.gurunars.item_list.Item

internal class ActionEdit<ItemType : Item>(
    private val itemConsumer: (item: ItemType) -> Unit
) : Action<ItemType> {

    override fun canPerform(
        all: IRoBox<List<ItemType>>,
        selectedItems: IRoBox<Set<ItemType>>
    ): IRoBox<Boolean> = selectedItems.oneWayBranch { size == 1 }

    override fun perform(
        all: List<ItemType>,
        selectedItems: Set<ItemType>,
        consumer: ItemSetChange<ItemType>
    ) = from {
        all.first { item -> selectedItems.indexOfFirst { it.id == item.id } != -1 }
    }.to { itemConsumer(it) }

}